src/main/java/me/desair/tus/server/upload/disk/CachedDiskAndLockingStorageService.java

```diff
@@ -0,0 +1,98 @@
+package me.desair.tus.server.upload.disk;
+
+import me.desair.tus.server.exception.TusException;
+import me.desair.tus.server.exception.UploadNotFoundException;
+import me.desair.tus.server.upload.UploadIdFactory;
+import me.desair.tus.server.upload.UploadInfo;
+import me.desair.tus.server.upload.UploadLock;
+import me.desair.tus.server.upload.UploadLockingService;
+import me.desair.tus.server.upload.UploadStorageService;
+
+import java.io.IOException;
+import java.lang.ref.WeakReference;
+import java.util.UUID;
+
+/**
+ * Combined implementation of {@link UploadStorageService} and {@link UploadLockingService} based on
+ * {@link DiskStorageService} but allowing to reduce disk operations during a request processing by caching UploadInfo
+ * in the memory.
+ * <p>
+ * UploadLockingService service is used as a delegate to cleanup cached data on releasing a lock.
+ */
+public class CachedDiskAndLockingStorageService extends DiskStorageService implements UploadLockingService {
+    private final ThreadLocal<WeakReference<UploadInfo>> uploadInfoCache = new ThreadLocal<>();
+    private final UploadLockingService lockingServiceDelegate;
+
+    public CachedDiskAndLockingStorageService(UploadIdFactory idFactory,
```

Comment:
This would change to `public CachedDiskAndLockingStorageService(UploadStorageService storageServiceDelegate, UploadLockingService lockingServiceDelegate)`